### DISCUSSION TITLE
Combining atoms with async generator

### DISCUSSION THREAD

**Discussion by AlexanderMoroz** [pmndrs/jotai#1917] [Q&A] [answered]
Hi everyone!

First of all - thanks for making such nice flexible library like 'jotai'! really enjoy working with it. 
I've faced an issue and would like to get any suggestion/ideas that might be helpful.

In our current project we use quite nice and pretty new library to work with graphql - `gqty` [https://gqty.dev/](url).
They just updated work with subscriptions, but with our complicated and nested structure it's not working as good as it should be, so I'm think about trying to implement my own way of working with their `core/subscription`

Main issue - gqty `subscribe` method returning async generator, and each `yield` contains latest data which needed to be set somewhere. I'm looking to create some kind of `atomWithSubscription` like jotai had `atomWithObservable` which is kind of what we need.

that's my half-finished thoughts process and were I'm a bit stuck 

```
const pinsSubscriptionAtom = atom(async (get, set) => {
  const assetId = get(assetIdAtom);
  const $data = subscribe(({ subscription: { items } }) => {
    const data = items();
    return data.map((item) => item.id);
  });

  for await (const data of $data) {
    // here I want to update current atom value
  }
});
```

so there is two issues:
 First one is to find appropriate place to call `set` method to set value
second one is to find correct place to `unsubscribe()` from subscription

for sure it can be done with useEffect and useState in react, but I'm curious is it possible to do on atom-like level

Would like to hear any suggestions/advices/criticism whatever! Thanks a lot!

**Comment by himself65** [comment, marked as answer]
I would suggest you use `rxjs` with `atom`

https://jotai.org/docs/utilities/async#atomwithobservable

```tsx
export const currentPageMetaAtom = atomWithObservable<PageMeta[]>(
  (get) => {
    return from(get(currentWorkspaceAtom)).pipe(
      filter((workspace): workspace is Workspace => workspace !== null),
      mergeMap(workspace => {
        const group = new DisposableGroup()
        return new Observable<PageMeta[]>((subscriber) => {
          const cleanup = pageAdded.on(() => {
            subscriber.next(workspace.meta.pageMetas)
          }))
          return () => {
            cleanup()
          }
        })
      })
    )
  })
```

**Reply by AlexanderMoroz** [reply]
yeah, I had an idea to create observable from generator and use atomWithObservable, only one downside is to bring rxjs dependency to the repo…
